container-storage-setup.sh

```diff
@@ -1109,6 +1110,84 @@ setup_extra_volume() {
   fi
 }
 
+# This is used only in compatibility mode. We are still using systemd
+# mount unit only for compatibility mode. Reason being that upon service
+# restart, we run into races and device might not yet be up when we try
+# to mount it. And we don't know if this is first time start and we need
+# to create volume or this is restart and we need to wait for device. So
+# continue to use systemd mount unit for compatibility mode.
+setup_systemd_mount_unit_compat() {
+  local filename
+  local vg=$1
+  local lv_name=$2
+  local mount_dir=$3
+  local unit_file_path
+
+  # filename must match the path ${mount_dir}.
+  # e.g if ${mount_dir} is /var/lib/containers
+  # then filename will be var-lib-containers.mount
+  filename=$(systemd_escaped_filename ${mount_dir})
+  unit_file_path="/etc/systemd/system/$filename"
+
+  # If unit file already exists, nothing to do.
+  [ -f "$unit_file_path" ] && return 0
+
+  cat <<EOF > "$unit_file_path"
+[Unit]
+Description=Mount $lv_name on $mount_dir directory.
+Before=docker-storage-setup.service
+
+[Mount]
+What=/dev/$vg/$lv_name
+Where=${mount_dir}
+Type=xfs
+Options=defaults
+
+[Install]
+WantedBy=docker-storage-setup.service
+EOF
+  systemctl enable $filename >/dev/null 2>&1
+  systemctl start $filename
+}
+
+setup_extra_lv_fs_compat() {
+  [ -z "$_RESOLVED_MOUNT_DIR_PATH" ] && return 0
+  if ! setup_extra_dir $_RESOLVED_MOUNT_DIR_PATH; then
+    return 1
+  fi
+  # If we are restarting, then extra volume should exist. This unit
+  # file is dependent on extra volume mount unit file. That means this
+  # code should run after mount unit has activated successfully. That
+  # means after extra volume has come up.
+
+  # We had got rid of this logic and reintroducing it back. That means
+  # there can be configurations out there which have extra volume but
+  # don't have unit file. So in such case, drop a unit file now. This
+  # is still racy though. There is no guarantee that volume will be
+  # up by the time this code runs when unit file is not present already.
+  if extra_volume_exists $CONTAINER_ROOT_LV_NAME $VG; then
+    if ! setup_systemd_mount_unit_compat "$VG" "$CONTAINER_ROOT_LV_NAME" "$_RESOLVED_MOUNT_DIR_PATH"; then
```

Comment:
If admin changes config file to start mounting on a different path, then a mount unit will be generated.  Ideally I am expecting user to "reset" existing configuration and then make changes to config file. Otherwise there are many possibilities and many things can go wrong.

On a side note, to cover for all these corner cases, I have written new commands for container-storage-setup and at some point of time we should make use of it. These commands store their
own metadata and don't allow arbitrary changes to existing configuration.